File: pkg/proxy/internal/adapters/http/vpc.go
package httphandlers

import (
	"log"
	"net/http"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
	"github.com/go-chi/chi/v5"
)

func (h *ProxyHandler) registerVpcRoutes(r chi.Router) {
	r.Route("/vpc", func(r chi.Router) {
		// VPCs
		r.Get("/vpcs", h.listVpcs)
		r.Post("/vpcs", h.createVpc)
		r.Delete("/vpcs/{vpcId}", h.deleteVpc)
		// Subnets
		r.Get("/subnets", h.listSubnets)
		r.Post("/subnets", h.createSubnet)
		r.Delete("/subnets/{subnetId}", h.deleteSubnet)
		// Route Tables
		r.Get("/route-tables", h.listRouteTables)
		r.Post("/route-tables", h.createRouteTable)
		r.Delete("/route-tables/{rtbId}", h.deleteRouteTable)
		r.Post("/route-tables/{rtbId}/routes", h.createVpcRoute)
		r.Delete("/route-tables/{rtbId}/routes/{cidr}", h.deleteVpcRoute)
		r.Post("/route-tables/{rtbId}/associate", h.associateRouteTable)
		r.Post("/route-tables/{rtbId}/disassociate", h.disassociateRouteTable)
		// Internet Gateways
		r.Get("/internet-gateways", h.listInternetGateways)
		r.Post("/internet-gateways", h.createInternetGateway)
		r.Delete("/internet-gateways/{igwId}", h.deleteInternetGateway)
		r.Post("/internet-gateways/{igwId}/attach", h.attachInternetGateway)
		r.Post("/internet-gateways/{igwId}/detach", h.detachInternetGateway)
		// NAT Gateways
		r.Get("/nat-gateways", h.listNatGateways)
		r.Post("/nat-gateways", h.createNatGateway)
		r.Delete("/nat-gateways/{natGwId}", h.deleteNatGateway)
		// Network ACLs
		r.Get("/network-acls", h.listNetworkAcls)
		r.Post("/network-acls", h.createNetworkAcl)
		r.Delete("/network-acls/{naclId}", h.deleteNetworkAcl)
		r.Post("/network-acls/{naclId}/entries", h.createNetworkAclEntry)
		r.Delete("/network-acls/{naclId}/entries/{ruleNumber}", h.deleteNetworkAclEntry)
		// VPC Flow Logs
		r.Get("/flow-logs", h.listFlowLogs)
		r.Post("/flow-logs", h.createFlowLogs)
		r.Delete("/flow-logs/{flowLogId}", h.deleteFlowLogs)
		// Elastic IPs
		r.Get("/elastic-ips", h.listElasticIps)
		r.Post("/elastic-ips", h.allocateElasticIp)
		r.Delete("/elastic-ips/{allocationId}", h.releaseElasticIp)
	})
}

// ---------------------------------------------------------------------------
// VPCs
// ---------------------------------------------------------------------------

func (h *ProxyHandler) listVpcs(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.DescribeVpcsInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().DescribeVpcs(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list VPCs", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createVpc(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.CreateVpcInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().CreateVpc(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create VPC", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteVpc(w http.ResponseWriter, r *http.Request) {
	vpcID := urlParam(r, "vpcId")
	input := &ec2.DeleteVpcInput{
		VpcId: aws.String(vpcID),
	}
	result, err := h.Svc.Vpc().DeleteVpc(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete VPC", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// ---------------------------------------------------------------------------
// Subnets
// ---------------------------------------------------------------------------

func (h *ProxyHandler) listSubnets(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.DescribeSubnetsInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().DescribeSubnets(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list subnets", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createSubnet(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.CreateSubnetInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().CreateSubnet(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create subnet", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteSubnet(w http.ResponseWriter, r *http.Request) {
	subnetID := urlParam(r, "subnetId")
	input := &ec2.DeleteSubnetInput{
		SubnetId: aws.String(subnetID),
	}
	result, err := h.Svc.Vpc().DeleteSubnet(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete subnet", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// ---------------------------------------------------------------------------
// Route Tables
// ---------------------------------------------------------------------------

func (h *ProxyHandler) listRouteTables(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.DescribeRouteTablesInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().DescribeRouteTables(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list route tables", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createRouteTable(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.CreateRouteTableInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().CreateRouteTable(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create route table", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteRouteTable(w http.ResponseWriter, r *http.Request) {
	rtbID := urlParam(r, "rtbId")
	input := &ec2.DeleteRouteTableInput{
		RouteTableId: aws.String(rtbID),
	}
	result, err := h.Svc.Vpc().DeleteRouteTable(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete route table", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createVpcRoute(w http.ResponseWriter, r *http.Request) {
	rtbID := urlParam(r, "rtbId")
	bodyBytes := readBody(r)
	var body struct {
		DestinationCidrBlock    *string `json:"DestinationCidrBlock"`
		GatewayId               *string `json:"GatewayId"`
		NatGatewayId            *string `json:"NatGatewayId"`
		NetworkInterfaceId      *string `json:"NetworkInterfaceId"`
		VpcPeeringConnectionId  *string `json:"VpcPeeringConnectionId"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &ec2.CreateRouteInput{
		RouteTableId:            aws.String(rtbID),
		DestinationCidrBlock:    body.DestinationCidrBlock,
		GatewayId:               body.GatewayId,
		NatGatewayId:            body.NatGatewayId,
		NetworkInterfaceId:      body.NetworkInterfaceId,
		VpcPeeringConnectionId:  body.VpcPeeringConnectionId,
	}
	result, err := h.Svc.Vpc().CreateRoute(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create route", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteVpcRoute(w http.ResponseWriter, r *http.Request) {
	rtbID := urlParam(r, "rtbId")
	cidr := urlParam(r, "cidr")
	input := &ec2.DeleteRouteInput{
		RouteTableId:         aws.String(rtbID),
		DestinationCidrBlock: aws.String(cidr),
	}
	result, err := h.Svc.Vpc().DeleteRoute(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete route", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) associateRouteTable(w http.ResponseWriter, r *http.Request) {
	rtbID := urlParam(r, "rtbId")
	bodyBytes := readBody(r)
	var body struct {
		SubnetId *string `json:"SubnetId"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &ec2.AssociateRouteTableInput{
		RouteTableId: aws.String(rtbID),
		SubnetId:     body.SubnetId,
	}
	result, err := h.Svc.Vpc().AssociateRouteTable(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to associate route table", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) disassociateRouteTable(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		AssociationId *string `json:"AssociationId"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &ec2.DisassociateRouteTableInput{
		AssociationId: body.AssociationId,
	}
	result, err := h.Svc.Vpc().DisassociateRouteTable(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to disassociate route table", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// ---------------------------------------------------------------------------
// Internet Gateways
// ---------------------------------------------------------------------------

func (h *ProxyHandler) listInternetGateways(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.DescribeInternetGatewaysInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().DescribeInternetGateways(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list internet gateways", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createInternetGateway(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.CreateInternetGatewayInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().CreateInternetGateway(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create internet gateway", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteInternetGateway(w http.ResponseWriter, r *http.Request) {
	igwID := urlParam(r, "igwId")
	input := &ec2.DeleteInternetGatewayInput{
		InternetGatewayId: aws.String(igwID),
	}
	result, err := h.Svc.Vpc().DeleteInternetGateway(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete internet gateway", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) attachInternetGateway(w http.ResponseWriter, r *http.Request) {
	igwID := urlParam(r, "igwId")
	bodyBytes := readBody(r)
	var body struct {
		VpcId *string `json:"VpcId"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &ec2.AttachInternetGatewayInput{
		InternetGatewayId: aws.String(igwID),
		VpcId:             body.VpcId,
	}
	result, err := h.Svc.Vpc().AttachInternetGateway(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to attach internet gateway", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) detachInternetGateway(w http.ResponseWriter, r *http.Request) {
	igwID := urlParam(r, "igwId")
	bodyBytes := readBody(r)
	var body struct {
		VpcId *string `json:"VpcId"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &ec2.DetachInternetGatewayInput{
		InternetGatewayId: aws.String(igwID),
		VpcId:             body.VpcId,
	}
	result, err := h.Svc.Vpc().DetachInternetGateway(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to detach internet gateway", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// ---------------------------------------------------------------------------
// NAT Gateways
// ---------------------------------------------------------------------------

func (h *ProxyHandler) listNatGateways(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.DescribeNatGatewaysInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().DescribeNatGateways(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list NAT gateways", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createNatGateway(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.CreateNatGatewayInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().CreateNatGateway(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create NAT gateway", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteNatGateway(w http.ResponseWriter, r *http.Request) {
	natGwID := urlParam(r, "natGwId")
	input := &ec2.DeleteNatGatewayInput{
		NatGatewayId: aws.String(natGwID),
	}
	result, err := h.Svc.Vpc().DeleteNatGateway(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete NAT gateway", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// ---------------------------------------------------------------------------
// Network ACLs
// ---------------------------------------------------------------------------

func (h *ProxyHandler) listNetworkAcls(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.DescribeNetworkAclsInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().DescribeNetworkAcls(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list network ACLs", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createNetworkAcl(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.CreateNetworkAclInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().CreateNetworkAcl(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create network ACL", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteNetworkAcl(w http.ResponseWriter, r *http.Request) {
	naclID := urlParam(r, "naclId")
	input := &ec2.DeleteNetworkAclInput{
		NetworkAclId: aws.String(naclID),
	}
	result, err := h.Svc.Vpc().DeleteNetworkAcl(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete network ACL", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createNetworkAclEntry(w http.ResponseWriter, r *http.Request) {
	naclID := urlParam(r, "naclId")
	bodyBytes := readBody(r)
	var body ec2.CreateNetworkAclEntryInput
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	body.NetworkAclId = aws.String(naclID)
	result, err := h.Svc.Vpc().CreateNetworkAclEntry(h.ctx, &body)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create network ACL entry", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteNetworkAclEntry(w http.ResponseWriter, r *http.Request) {
	naclID := urlParam(r, "naclId")
	ruleNumber := urlParam(r, "ruleNumber")
	input := &ec2.DeleteNetworkAclEntryInput{
		NetworkAclId: aws.String(naclID),
		RuleNumber:   aws.Int32(parseInt32(ruleNumber)),
		Egress:       aws.Bool(false),
	}
	result, err := h.Svc.Vpc().DeleteNetworkAclEntry(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to delete network ACL entry", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// ---------------------------------------------------------------------------
// VPC Flow Logs
// ---------------------------------------------------------------------------

func (h *ProxyHandler) listFlowLogs(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.DescribeFlowLogsInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().DescribeFlowLogs(h.ctx, input)
	if err != nil {
		if isUnsupportedError(err) {
			log.Printf("Flow Logs not supported by emulator: %v", err)
			writeJSON(w, http.StatusOK, map[string]interface{}{"FlowLogs": []interface{}{}, "Unsupported": true})
			return
		}
		sendError(w, http.StatusInternalServerError, "Failed to list flow logs", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createFlowLogs(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.CreateFlowLogsInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().CreateFlowLogs(h.ctx, input)
	if err != nil {
		if isUnsupportedError(err) {
			log.Printf("Flow Logs not supported by emulator: %v", err)
			writeJSON(w, http.StatusOK, map[string]interface{}{"Unsupported": true})
			return
		}
		sendError(w, http.StatusInternalServerError, "Failed to create flow logs", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteFlowLogs(w http.ResponseWriter, r *http.Request) {
	flowLogID := urlParam(r, "flowLogId")
	input := &ec2.DeleteFlowLogsInput{
		FlowLogIds: []string{flowLogID},
	}
	result, err := h.Svc.Vpc().DeleteFlowLogs(h.ctx, input)
	if err != nil {
		if isUnsupportedError(err) {
			log.Printf("Flow Logs not supported by emulator: %v", err)
			writeJSON(w, http.StatusOK, map[string]interface{}{"Unsupported": true})
			return
		}
		sendError(w, http.StatusInternalServerError, "Failed to delete flow logs", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// ---------------------------------------------------------------------------
// Elastic IPs
// ---------------------------------------------------------------------------

func (h *ProxyHandler) listElasticIps(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.DescribeAddressesInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().DescribeAddresses(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list Elastic IPs", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) allocateElasticIp(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &ec2.AllocateAddressInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.Vpc().AllocateAddress(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to allocate Elastic IP", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) releaseElasticIp(w http.ResponseWriter, r *http.Request) {
	allocationID := urlParam(r, "allocationId")
	input := &ec2.ReleaseAddressInput{
		AllocationId: aws.String(allocationID),
	}
	result, err := h.Svc.Vpc().ReleaseAddress(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to release Elastic IP", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// ---------------------------------------------------------------------------
// parseInt32 helper
// ---------------------------------------------------------------------------

func parseInt32(s string) int32 {
	var i int32
	for _, c := range s {
		if c >= '0' && c <= '9' {
			i = i*10 + (int32(c) - '0')
		} else {
			break
		}
	}
	return i
}

File: pkg/ui/src/api/services/vpc.ts
/**
 * VPC Service API Client
 * REST client for VPC via Go proxy
 * @module api/services/vpc
 */

import { PROXY_BACKEND } from '@/config'
import { APIError } from '../client'
import type { EC2Vpc, EC2Subnet, EC2RouteTable, EC2InternetGateway, EC2NatGateway, EC2NetworkAcl, EC2FlowLog, EC2ElasticIp } from '@/api/types/aws'

const BASE_URL = `${PROXY_BACKEND.replace(/\/$/, '')}/vpc`

async function request<T>(method: string, path: string, body?: object): Promise<T> {
  const url = `${BASE_URL}${path}`
  const fetchOptions: RequestInit = { method }

  if (body !== undefined) {
    fetchOptions.headers = { 'Content-Type': 'application/json' }
    fetchOptions.body = JSON.stringify(body)
  }

  try {
    const response = await fetch(url, fetchOptions)
    const responseText = await response.text()

    if (!response.ok) {
      throw new APIError(`VPC request failed: ${responseText}`, response.status, 'vpc')
    }

    if (!responseText) {
      return {} as T
    }

    return JSON.parse(responseText) as T
  } catch (error) {
    if (error instanceof APIError) throw error
    console.error('VPC request error:', error)
    throw new APIError(`Failed to ${method} ${path}`, 500, 'vpc')
  }
}

// VPCs
export interface DescribeVpcsResponse {
  Vpcs: EC2Vpc[]
}

export async function describeVpcs(): Promise<DescribeVpcsResponse> {
  return request<DescribeVpcsResponse>('GET', '/vpcs')
}

export async function createVpc(params: { CidrBlock: string }): Promise<EC2Vpc> {
  return request<EC2Vpc>('POST', '/vpcs', params)
}

export async function deleteVpc(vpcId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/vpcs/${encodeURIComponent(vpcId)}`)
}

// Subnets
export interface DescribeSubnetsResponse {
  Subnets: EC2Subnet[]
}

export async function describeSubnets(): Promise<DescribeSubnetsResponse> {
  return request<DescribeSubnetsResponse>('GET', '/subnets')
}

export async function createSubnet(params: { VpcId: string; CidrBlock: string; AvailabilityZone?: string }): Promise<EC2Subnet> {
  return request<EC2Subnet>('POST', '/subnets', params)
}

export async function deleteSubnet(subnetId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/subnets/${encodeURIComponent(subnetId)}`)
}

// Route Tables
export interface DescribeRouteTablesResponse {
  RouteTables: EC2RouteTable[]
}

export async function describeRouteTables(): Promise<DescribeRouteTablesResponse> {
  return request<DescribeRouteTablesResponse>('GET', '/route-tables')
}

export async function createRouteTable(params: { VpcId: string }): Promise<EC2RouteTable> {
  return request<EC2RouteTable>('POST', '/route-tables', params)
}

export async function deleteRouteTable(rtbId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/route-tables/${encodeURIComponent(rtbId)}`)
}

export async function createRoute(rtbId: string, params: { DestinationCidrBlock: string; GatewayId?: string; NatGatewayId?: string }): Promise<Record<string, never>> {
  return request<Record<string, never>>('POST', `/route-tables/${encodeURIComponent(rtbId)}/routes`, params)
}

export async function deleteRoute(rtbId: string, cidr: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/route-tables/${encodeURIComponent(rtbId)}/routes/${encodeURIComponent(cidr)}`)
}

export async function associateRouteTable(rtbId: string, subnetId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('POST', `/route-tables/${encodeURIComponent(rtbId)}/associate`, { SubnetId: subnetId })
}

export async function disassociateRouteTable(associationId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('POST', `/route-tables/${encodeURIComponent(associationId)}/disassociate`)
}

// Internet Gateways
export interface DescribeInternetGatewaysResponse {
  InternetGateways: EC2InternetGateway[]
}

export async function describeInternetGateways(): Promise<DescribeInternetGatewaysResponse> {
  return request<DescribeInternetGatewaysResponse>('GET', '/internet-gateways')
}

export async function createInternetGateway(): Promise<EC2InternetGateway> {
  return request<EC2InternetGateway>('POST', '/internet-gateways')
}

export async function deleteInternetGateway(igwId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/internet-gateways/${encodeURIComponent(igwId)}`)
}

export async function attachInternetGateway(igwId: string, vpcId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('POST', `/internet-gateways/${encodeURIComponent(igwId)}/attach`, { VpcId: vpcId })
}

export async function detachInternetGateway(igwId: string, vpcId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('POST', `/internet-gateways/${encodeURIComponent(igwId)}/detach`, { VpcId: vpcId })
}

// NAT Gateways
export interface DescribeNatGatewaysResponse {
  NatGateways: EC2NatGateway[]
}

export async function describeNatGateways(): Promise<DescribeNatGatewaysResponse> {
  return request<DescribeNatGatewaysResponse>('GET', '/nat-gateways')
}

export async function createNatGateway(params: { SubnetId: string; AllocationId: string }): Promise<EC2NatGateway> {
  return request<EC2NatGateway>('POST', '/nat-gateways', params)
}

export async function deleteNatGateway(natGwId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/nat-gateways/${encodeURIComponent(natGwId)}`)
}

// Network ACLs
export interface DescribeNetworkAclsResponse {
  NetworkAcls: EC2NetworkAcl[]
}

export async function describeNetworkAcls(): Promise<DescribeNetworkAclsResponse> {
  return request<DescribeNetworkAclsResponse>('GET', '/network-acls')
}

export async function createNetworkAcl(params: { VpcId: string }): Promise<EC2NetworkAcl> {
  return request<EC2NetworkAcl>('POST', '/network-acls', params)
}

export async function deleteNetworkAcl(naclId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/network-acls/${encodeURIComponent(naclId)}`)
}

export async function createNetworkAclEntry(naclId: string, params: {
  RuleNumber: number
  Protocol: string
  PortRange?: { From: number; To: number }
  CidrBlock: string
  Egress: boolean
  RuleAction: 'allow' | 'deny'
}): Promise<Record<string, never>> {
  return request<Record<string, never>>('POST', `/network-acls/${encodeURIComponent(naclId)}/entries`, params)
}

export async function deleteNetworkAclEntry(naclId: string, ruleNumber: number): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/network-acls/${encodeURIComponent(naclId)}/entries/${ruleNumber}`)
}

// Flow Logs
export interface DescribeFlowLogsResponse {
  FlowLogs: EC2FlowLog[]
  Unsupported?: boolean
}

export async function describeFlowLogs(): Promise<DescribeFlowLogsResponse> {
  return request<DescribeFlowLogsResponse>('GET', '/flow-logs')
}

export async function createFlowLogs(params: {
  ResourceId: string
  LogDestinationType: string
  LogDestination: string
  TrafficType: string
}): Promise<EC2FlowLog> {
  return request<EC2FlowLog>('POST', '/flow-logs', params)
}

export async function deleteFlowLogs(flowLogId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/flow-logs/${encodeURIComponent(flowLogId)}`)
}

// Elastic IPs
export interface DescribeAddressesResponse {
  Addresses: EC2ElasticIp[]
}

export async function describeAddresses(): Promise<DescribeAddressesResponse> {
  return request<DescribeAddressesResponse>('GET', '/elastic-ips')
}

export async function allocateElasticIp(): Promise<EC2ElasticIp> {
  return request<EC2ElasticIp>('POST', '/elastic-ips')
}

export async function releaseElasticIp(allocationId: string): Promise<Record<string, never>> {
  return request<Record<string, never>>('DELETE', `/elastic-ips/${encodeURIComponent(allocationId)}`)
}

export default {
  describeVpcs,
  createVpc,
  deleteVpc,
  describeSubnets,
  createSubnet,
  deleteSubnet,
  describeRouteTables,
  createRouteTable,
  deleteRouteTable,
  createRoute,
  deleteRoute,
  associateRouteTable,
  disassociateRouteTable,
  describeInternetGateways,
  createInternetGateway,
  deleteInternetGateway,
  attachInternetGateway,
  detachInternetGateway,
  describeNatGateways,
  createNatGateway,
  deleteNatGateway,
  describeNetworkAcls,
  createNetworkAcl,
  deleteNetworkAcl,
  createNetworkAclEntry,
  deleteNetworkAclEntry,
  describeFlowLogs,
  createFlowLogs,
  deleteFlowLogs,
  describeAddresses,
  allocateElasticIp,
  releaseElasticIp,
}
